crates/oxc_semantic/src/counter.rs

```diff
@@ -19,6 +23,27 @@ pub(crate) struct Counts {
     pub references: usize,
 }
 
+impl Counts {
+    pub fn count(program: &Program) -> Self {
+        let mut counts = Counts::default();
+        counts.visit_program(program);
+        counts
+    }
+
+    pub fn assert_accurate(actual: &Self, estimated: &Self) {
+        assert_eq!(actual.nodes, estimated.nodes, "nodes count mismatch");
+        assert_eq!(actual.scopes, estimated.scopes, "scopes count mismatch");
+        assert_eq!(actual.references, estimated.references, "references count mismatch");
+        // `Counts` may overestimate number of symbols, because multiple `BindingIdentifier`s
+        // can result in only a single symbol.
+        // e.g. `var x; var x;` = 2 x `BindingIdentifier` but 1 x symbol.
+        // This is not a big problem - allocating a `Vec` with excess capacity is cheap.
+        // It's allocating with *not enough* capacity which is costly, as then the `Vec`
+        // will grow and reallocate.
+        assert_le!(actual.symbols, estimated.symbols, "symbols count mismatch");
```

Comment:
This can be written as `assert!(actual.symbols < estimated.symbols, "{} < {}", actual.symbols, estimated.symbols)`